examples/improved_wgan.py

```diff
@@ -0,0 +1,265 @@
+from keras.layers.merge import _Merge
+from keras.layers.convolutional import Convolution2D, AveragePooling2D, MaxPooling2D, Conv2DTranspose
+from keras.layers.normalization import BatchNormalization
+from keras.layers.advanced_activations import LeakyReLU
+from keras.optimizers import Adam
+from keras.datasets import mnist
+from keras import backend as K
+from functools import partial
+import tensorflow as tf
+
+try:
+    from PIL import Image
+except ImportError:
+    print('This script depends on pillow! Please install it (e.g. with pip install pillow)')
+    exit()
+
+BATCH_SIZE = 64
+TRAINING_RATIO = 5  # The training ratio is the number of discriminator updates per generator update. The paper uses 5.
+GRADIENT_PENALTY_WEIGHT = 10  # As per the paper
+
+
+def wasserstein_loss(y_true, y_pred):
+    """Calculates the Wasserstein loss for a sample batch.
+
+    The Wasserstein loss function is very simple to calculate. In a standard GAN, the discriminator
+    has a sigmoid output, representing the probability that samples are real or generated. In Wasserstein
+    GANs, however, the output is linear with no activation function! Instead of being constrained to [0, 1],
+    the discriminator wants to make the distance between its output for real and generated samples as large as possible.
+
+    The most natural way to achieve this is to label generated samples -1 and real samples 1, instead of the
+    0 and 1 used in normal GANs, so that multiplying the outputs by the labels will give you the loss immediately.
+
+    Note that the nature of this loss means that it can be (and frequently will be) less than 0."""
+    return K.mean(y_true * y_pred)
+
+
+def gradient_penalty_loss(y_true, y_pred, trainable_weights):
+    """Calculates the gradient penalty loss for a batch of "averaged" samples.
+
+    In Improved WGANs, the 1-Lipschitz constraint is enforced by adding a term to the loss function
+    that penalizes the network if the gradient norm moves away from 1. However, it is impossible to evaluate
+    this function at all points in the input space. The compromise used in the paper is to choose random points
+    on the lines between real and generated samples, and check the gradients at these points.
+
+    In order to evaluate the gradients, we must first run samples through the discriminator and evaluate the loss.
+    Then we get the gradients of the trainable discriminator weights, flatten the gradient tensors
+    and concatenate them into one 1D gradient tensor. The l2 norm and penalty can then be calculated from it.
+
+    Note that this loss function requires the model's trainable weights as input, but Keras only supports passing
+    y_true and y_pred to loss functions. To get around this, we make a partial() of the function with the
+    trainable_weights argument, and use that for model training."""
+    averaged_samples_loss = K.mean(y_pred)
+    gradients = K.gradients(averaged_samples_loss, trainable_weights)

```

Comment:
I double-checked your workings and you're right, I was taking the gradient of the discriminator w.r.t. the weights, not the inputs. Lesson learned there, I guess: Just because empirical performance is good doesn't mean you haven't left a nasty maths error somewhere in the code. Performance with the change seems roughly similar, possibly slightly faster convergence?